A warning for unsafe field uses

parsonsmatt opened the pull request:
Currently, a `WARNING` or `DEPRECATED` pragma on a record field label triggers for any use of a field label -- accessing a value, creating a record, record update, pattern matches, and named field puns. I propose that GHC add a mechanism whereby *unsafe* usages of record labels trigger warnings, and safe usages do not. Safe usages include record creation, record pattern matches, ``NamedFieldPuns``, and ``RecordWildCards``.

[Rendered](https://github.com/parsonsmatt/ghc-proposals/blob/unsafe-field-uses/proposals/0000-unsafe-field-uses.rst)

theindigamer commented:
Isn't this partially remedied by [`-Wpartial-fields`](https://downloads.haskell.org/~ghc/latest/docs/html/users_guide/using-warnings.html#ghc-flag--Wpartial-fields)? This emits a warning at the definition site.

cleverca22 commented:
named field puns allow you to pattern match on both the subtype in a sum type, and the fields by name, so its always safe

gridaphobe commented:
Another alternative might be to introduce a variant of the flag in #160, which would disable generation of **partial** field selector functions. I'm not sure how this would interact with record updates though, I suspect we would still need to address partial updates separately.

parsonsmatt commented:
> Isn't this partially remedied by -Wpartial-fields? This emits a warning at the definition site, not at the usage site though. @theindigamer 

As you noted, `-Wpartial-fields` provided a warning whenever you define partial fields. The proposal suggests a way to warn only on *unsafe* uses of partial fields. They are orthogonal concerns. Indeed, if you only permit safe uses of partial fields, there's no reason to warn when defining them.

> I'm not sure how this would interact with record updates though, I suspect we would still need to address partial updates separately. @gridaphobe 

Exactly -- the unsafe operations are record update and field access. Not generating top-level functions at all only solves half of the problem.

treeowl commented:
@gridaphobe, I've never actually thought about what update does when an updated field is missing. I would intuitively want the missing field to be ignored.

```haskell
data Foo
    = Bar {bar :: Char, bar' :: Int}
    | Baz {bar :: Char, baz :: String}
```

I would expect

```haskell
(Bar 'a' 3){bar = 'b', baz = "hi"} = Bar 'b' 3
```

If that's not how it works, maybe it should.

gridaphobe commented:
Currently, you get a runtime error.

```
ghci> (Bar 'a' 3){bar = 'b', baz = "hi"}
*** Exception: <interactive>:8:1-34: Non-exhaustive patterns in record update
```

And the error is a bit confusing at that. This doesn't really look like a non-exhaustive pattern match. I'd expect the error to be something along the lines of "Couldn't update missing 'baz' field".

I don't think silently ignoring the missing field is a good behavior, though perhaps I could be convinced otherwise. My concern is that doing so would just punt the error downstream where it will be harder to debug. But on the other hand, a downstream function can't expect `Bar` to have a `baz` field in the first place..

dcoutts commented:
I'd argue the `WARNING` and `DEPRECATED` pragmas are doing the right thing, and this use case wants a new `-Wxxx` flag, that does what you're asking for, unrelated to the `WARNING` and `DEPRECATED` pragmas.

gridaphobe commented:
@treeowl yes, but people have different expectations of `Map`s and records. A `Map` does not guarantee the presence of any key whereas a record does (barring the present issue of partial fields, which is often considered a misfeature).

In my mind, the more convincing argument for your preferred behavior is that a consumer of a `Bar` record cannot expect it to have a `baz` field in the first place, and so cannot be surprised that the update had no effect. They could be surprised that they got a `Bar` record instead of a `Baz`, but that's a problem regardless of the field-update behavior!

simonpj commented:
Can you be more precise in your "proposed change specification".  In 2.2 you say only things like "When this flag is enabled, GHC will issue a warning for any unsafe usage of a record selector".  But what (precisely) is an "unsafe usage"?  Similarly in 2.1. 

Can you say precisely what new warnings will be issued that are not currently warned about?

Also as @partsonsmatt says above, his example `r4` is a definite error.  Do we fail on tat at the moment?

ndmitchell commented:
I wonder if what you're really trying to get at is best thought of as two separate ideas:

* Marking the generated selector functions as PARTIAL if they are partial.
* Warning when PARTIAL functions are used.

Now the PARTIAL concept doesn't have to be separately expressible, but that lets you talk about the two concepts separately, and I think is what you are saying.

parsonsmatt commented:
@simonpj I have updated the proposal with a list of safe and unsafe uses of record fields that I'm aware of, along with an example of how the flag would cause warnings.

@ndmitchell That covers warning when using them as accessors, but not for update syntax (equally unsafe). Currently, we either a) have partial record fields and try real hard not to use them unsafely, or b) warn on any use of a partial record field via the `WARNING` pragma, which forbids safe use as well. The proposal aims to allow users to use partial fields safely and receive compile-time warnings for unsafe use.

ndmitchell commented:
Thanks @parsonsmatt - you're right. Reading the proposal the use of a pragma for warnings seems a bit odd - why not ditch the pragma from the proposal and put it in as an alternative? It feels like you are arguing for two things. Or do you feel that you need both and they complement each other?

parsonsmatt commented:
@ndmitchell They are meant to be alternative solutions to the same problem. I would only want to implement one of them. I'll reword the intro to that section to make that more clear.

simonpj commented:
> @simonpj I have updated the proposal with a list of safe and unsafe uses of record fields that I'm aware of, along with an example of how the flag would cause warnings.

Thanks.  But I'm still very lost.  (And something funny has happened to the typesetting of the examples.)  In 2.1 you say

> This would then issue a warning in any unsafe usage of a field, but safe usage would be permitted.

What is an "unsafe usage of a field"?

In 2.2 you say

> An unsafe use of a partial flag is one which may fail at runtime. These are listed here.

What is a "partial flag" and how might it fail at runtime?

Let me see if I can guess.  There are precisely three way in which a record field `wuggle` can be used in Haskell

1.  In pattern matching.  The constructor `K` is explicit, so it is easy for GHC to error if `wuggle` is not a field of `K`:
```
f (K { wuggle = pat }) = blah
```
2.  In record construction.  Again the constructor `K` is explicit:
```
f x = K { wuggle = 3 }
```
3.  In record updates.  Here the constructor is not epxlicit
```
f x = x { wuggle= 3 }
```
4.  In a record selector.  These are ordinary top-level functions, auto-generated from the data type
```
wuggle (K _ _ v _) = v
```
Now, I think that you are not proposing anything new for (1) or (2).  Correct?  So it affects only (3) and (4)?  

For (3) under precisely what circumstances to you expect a record update to emit an error or warning.  I think:
* Error if no data constructor has all the fields that are being updated.
* Warning if any constructor lacks all the updated fields.

For (4) I think you propose that every occurrence of a partial record selector function be warned about?

parsonsmatt commented:
Thank you for your patience @simonpj :smile: I've updated the proposal to address these concerns and the formatting issue. To directly answer:

> What is an "unsafe usage of a field"?

I've moved my definition for this up to the first paragraph: " Safe usages are record creation and record pattern matches (either via the standard syntax or NamedFieldPuns and RecordWildCards). Unsafe uses are record update and accessor functions."

> What is a "partial flag" and how might it fail at runtime?

A typo! I mean "partial field" and have corrected the proposal.

> Let me see if I can guess. There are precisely three way in which a record field wuggle can be used in Haskell (snip)

Your guess is exactly right :)

> Now, I think that you are not proposing anything new for (1) or (2). Correct? So it affects only (3) and (4)?

That is correct -- these are currently totally safe, and GHC will error if you attempt to use a field that doesn't belong to the constructor. 

> For (3) under precisely what circumstances to you expect a record update to emit an error or warning. I think:
> 
> * Error if no data constructor has all the fields that are being updated.
> * Warning if any constructor lacks all the updated fields.

Yup! I hadn't considered the first bullet, but that would be good.

> For (4) I think you propose that every occurrence of a partial record selector function be warned about?

Yes, that is correct.

simonpj commented:
Thanks for revising.  Alas, I still find it hard to understand.  For example "An unsafe use of a partial field is one which may fail at runtime."  But you haven't defined "partial field".

Why not list the four cases I listed, with a concrete example of each.  And say that your proposal simply does not affect (1) and (2).  And list the changes to (3) and (4) more or less as above, again with a concrete example in every case.

I don't think it makes sense to speak of an "unsafe use of a field `wug`".   For record update syntax, is
```
e { wug = True, blog = 4 }
```
an unsafe use of `wug` or `blob` or both?  Remember, `wug` might even be present in every constructor.    Also you want to *warn* in some cases and *error* in others, so there are degrees of unsafeness.

Much better not to introduce this confusing vocabulary in the first place.  Now that I understand your proposal better I think it is precisely:

* Warn about every use of `wug` as a record selector, if `wug` does not appear in every constructor of the type
* Error on every record update `e { fld1 = e1, .. fldn = en }`, if there is no constructor with all the fields `fld1`..`fldn`.
* Warn about every record update `e { fld1 = e1, .. fldn = en }` in which there is at least one data constructor that does not have all of `fld1`..`fldn`.

Is that it?  Does named field puns etc interact at all with this?

You could be very precise.  "Definition: suppose a data type `T` defines a field `wug`.  An "unsafe use of `wug`" is

adamgundry commented:
Making this a warning that affects use sites seems like the right thing to me. Being able to choose at definition sites whether or not to warn for particular fields doesn't seem terribly useful, and if I find myself using a third-party library that defines partial fields, I might well want to be warned!

The proposal doesn't currently mention the `HasField` magic class, which provides access to selector functions (and hopefully soon update functions, per #158). I think solving a `HasField` constraint involving a partial field should emit a warning, because it is essentially equivalent to using a selector function.

nomeata commented:
Little activity recently, marking this as dormant.

parsonsmatt commented:
> "Is that it? Does named field puns etc interact at all with this?" @simonpj 

Yes, that's precisely what I mean.

nomeata commented:
Closing this due to inactivity, but it can be reopened when necessary.

nomeata commented:
Indeed
